File: sampledata/benchmarks/pypdfsuit/bench.py
#!/usr/bin/env python3

import math
import multiprocessing
import sys
import time
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(REPO_ROOT / "sampledata" / "gopdflib" / "zerodha"))
sys.path.insert(0, str(REPO_ROOT / "bindings" / "python"))

from pypdfsuit import generate_pdf  # noqa: E402
from pypdfsuit_bench import build_retail_template  # noqa: E402

ITERATIONS = 10
NUM_WORKERS = min(48, ITERATIONS)


def run_once(run_index):
    """Each worker builds its own template and generates the PDF independently."""
    start = time.perf_counter()
    generate_pdf(build_retail_template())
    return (time.perf_counter() - start) * 1000


def main() -> None:
    print("=== PyPDFSuit Concurrent Zerodha Benchmark ===")
    print(f"Iterations: {ITERATIONS} | Workers: {NUM_WORKERS}")

    total_start = time.perf_counter()
    with multiprocessing.Pool(processes=NUM_WORKERS) as pool:
        all_times = pool.map(run_once, range(1, ITERATIONS + 1))
    total_seconds = time.perf_counter() - total_start

    for i, elapsed in enumerate(all_times, start=1):
        print(f"Run {i}: {elapsed:.2f} ms")

    sorted_times = sorted(all_times)
    p95_idx = max(0, math.ceil(len(sorted_times) * 0.95) - 1)

    print()
    print(f"Min: {min(all_times):.2f} ms")
    print(f"Avg: {sum(all_times)/len(all_times):.2f} ms")
    print(f"P95: {sorted_times[p95_idx]:.2f} ms")
    print(f"Max: {max(all_times):.2f} ms")
    print(f"Throughput: {ITERATIONS / total_seconds:.2f} ops/sec")


if __name__ == "__main__":
    main()

File: sampledata/benchmarks/jspdf/bench.js
'use strict';

const { jsPDF } = require('jspdf');
const fs = require('fs');
const path = require('path');
const { performance } = require('perf_hooks');
const { Worker, isMainThread, workerData, parentPort } = require('worker_threads');

const DATA_PATH = path.resolve(__dirname, '../data.json');
const data = JSON.parse(fs.readFileSync(DATA_PATH, 'utf8'));
const iterations = 10;
const NUM_WORKERS = Math.min(48, iterations);

let autoTable = null;
try {
    autoTable = require('jspdf-autotable');
} catch (e) {
    // optional dependency
}

function buildBody() {
    return data.map((record) => [
        String(record.id),
        record.name,
        record.email,
        record.role,
        record.desc,
    ]);
}

function drawTextFallback(doc) {
    let y = 30;
    const pageHeight = doc.internal.pageSize.height;
    doc.setFontSize(10);
    for (const r of data) {
        if (y > pageHeight - 20) {
            doc.addPage();
            y = 20;
        }
        doc.text(`${r.id} | ${r.name} | ${r.email} | ${r.role}`, 14, y);
        y += 10;
    }
}

function runOnce() {
    const start = performance.now();
    const doc = new jsPDF({ putOnlyUsedFonts: true });
    doc.text('User Report', 14, 20);

    const body = buildBody();
    if (doc.autoTable) {
        doc.autoTable({ startY: 25, head: [['ID', 'Name', 'Email', 'Role', 'Description']], body });
    } else if (autoTable) {
        try {
            const fn = typeof autoTable === 'function' ? autoTable
                : (autoTable.default && typeof autoTable.default === 'function') ? autoTable.default
                : null;
            if (fn) fn(doc, { startY: 25, head: [['ID', 'Name', 'Email', 'Role', 'Description']], body });
            else drawTextFallback(doc);
        } catch (e) {
            drawTextFallback(doc);
        }
    } else {
        drawTextFallback(doc);
    }

    doc.output('arraybuffer'); // Generate bytes in-memory — no disk I/O in hot path
    return performance.now() - start;
}

if (isMainThread) {
    function spawnWorker(runIndex) {
        return new Promise((resolve, reject) => {
            const worker = new Worker(__filename, { workerData: { runIndex } });
            worker.on('message', (msg) => resolve(msg));
            worker.on('error', reject);
            worker.on('exit', (code) => {
                if (code !== 0) reject(new Error(`Worker ${runIndex} exited with code ${code}`));
            });
        });
    }

    async function main() {
        console.log('=== jsPDF Data Benchmark ===');
        console.log(`Iterations: ${iterations} | Workers: ${NUM_WORKERS}`);

        const totalStart = performance.now();
        const results = await Promise.all(
            Array.from({ length: iterations }, (_, i) => spawnWorker(i + 1))
        );
        const totalSeconds = (performance.now() - totalStart) / 1000;

        // Print runs in order
        const ordered = results.sort((a, b) => a.runIndex - b.runIndex);
        for (const { runIndex, elapsed } of ordered) {
            console.log(`Run ${runIndex}: ${elapsed.toFixed(2)} ms`);
        }

        const timings = ordered.map((r) => r.elapsed);
        const sorted = [...timings].sort((a, b) => a - b);
        const p95idx = Math.max(0, Math.ceil(sorted.length * 0.95) - 1);

        console.log('');
        console.log(`Min:        ${sorted[0].toFixed(2)} ms`);
        console.log(`Avg:        ${(sorted.reduce((s, v) => s + v, 0) / sorted.length).toFixed(2)} ms`);
        console.log(`P95:        ${sorted[p95idx].toFixed(2)} ms`);
        console.log(`Max:        ${sorted[sorted.length - 1].toFixed(2)} ms`);
        console.log(`Throughput: ${(timings.length / totalSeconds).toFixed(2)} ops/sec`);
    }

    main().catch((err) => { console.error(err); process.exit(1); });
} else {
    // Worker thread: generate one PDF and report elapsed time
    const elapsed = runOnce();
    parentPort.postMessage({ elapsed, runIndex: workerData.runIndex });
}
